File: training-data/erlang/rabin_karp_search.erl
-module(rabin_karp_search).
-export([search/2]).

%% A simplified Rabin-Karp search: instead of maintaining a rolling
%% hash incrementally, each candidate window is hashed and compared
%% by hash first, then confirmed with an exact match to rule out hash
%% collisions, before sliding one position to the right.

search(Text, Pattern) ->
    N = length(Text),
    M = length(Pattern),
    search(Text, Pattern, hash(Pattern), N, M, 0).

search(_Text, _Pattern, _PatternHash, N, M, Start) when Start + M > N ->
    not_found;
search(Text, Pattern, PatternHash, N, M, Start) ->
    Window = lists:sublist(Text, Start + 1, M),
    case hash(Window) =:= PatternHash andalso Window =:= Pattern of
        true -> {found, Start};
        false -> search(Text, Pattern, PatternHash, N, M, Start + 1)
    end.

hash(Str) ->
    lists:foldl(fun(C, Acc) -> (Acc * 256 + C) rem 1000000007 end, 0, Str).

run() ->
    io:format("~p~n", [search("abxabcabcaby", "abcaby")]),
    io:format("~p~n", [search("hello world", "xyz")]).

File: training-data/elixir/rabin_karp_search.exs
defmodule RabinKarp do
  @base 256
  @modulus 101

  # Finds all starting indices where `pattern` occurs in `text`, using a
  # rolling hash as a cheap pre-filter before confirming an exact match.
  def search(text, pattern) do
    n = String.length(text)
    m = String.length(pattern)

    if m > n do
      []
    else
      pattern_hash = pattern |> String.to_charlist() |> hash()

      Enum.filter(0..(n - m), fn start ->
        window = String.slice(text, start, m)
        hash(String.to_charlist(window)) == pattern_hash and window == pattern
      end)
    end
  end

  defp hash(chars), do: Enum.reduce(chars, 0, fn c, acc -> rem(acc * @base + c, @modulus) end)
end

IO.inspect(RabinKarp.search("abracadabra", "abra"))
IO.inspect(RabinKarp.search("aaaaa", "aa"))
IO.inspect(RabinKarp.search("hello", "xyz"))
